packages/node/src/login/oidc/refresh/TokenRefresher.ts

```diff
@@ -42,21 +50,65 @@ export interface ITokenRefresher {
 export default class TokenRefresher implements ITokenRefresher {
   constructor(
     @inject("storageUtility") private storageUtility: IStorageUtility,
-    @inject("tokenRequester") private tokenRequester: ITokenRequester
+    @inject("issuerConfigFetcher")
+    private issuerConfigFetcher: IIssuerConfigFetcher,
+    @inject("clientRegistrar") private clientRegistrar: IClientRegistrar
   ) {}
 
-  async refresh(localUserId: string): Promise<void> {
-    // Get the refresh token and the issuer
-    const refreshToken = await this.storageUtility.getForUser(
-      localUserId,
-      "refreshToken",
-      { errorIfNull: true, secure: true }
+  async refresh(
+    sessionId: string,
+    refreshToken?: string,
+    dpopKey?: JWK.ECKey
+  ): Promise<TokenSet> {
+    const oidcContext = await loadOidcContextFromStorage(
+      sessionId,
+      this.storageUtility,
+      this.issuerConfigFetcher
     );
-    /* eslint-disable camelcase */
-    await this.tokenRequester.request(localUserId, {
-      grant_type: "refresh_token",
-      refresh_token: refreshToken as string, // This cast can be safely made because getForUser will error if refreshToken is null
+    const issuer = new Issuer(configToIssuerMetadata(oidcContext.issuerConfig));
+    // This should also retrieve the client from storage
+    const clientInfo: IClient = await this.clientRegistrar.getClient(
+      { sessionId },
+      oidcContext.issuerConfig
+    );
+    const client = new issuer.Client({
+      client_id: clientInfo.clientId,
+      client_secret: clientInfo.clientSecret,
+    });
+
+    if (refreshToken === undefined) {
```

Comment:
This conditional seems weirdly placed here - i.e. its seems to be testing the value of the input param, so should it just be the first line of the function? And why is the input param optional if we immediately throw if it's `undefined`?

Reply:
Originally, the logic was that the refresh token could be available in storage, and not necessarily provided as a parameter: that's why it is optional. You are right that the test should be at the top of the function either way (if the token cannot be retrieved from storage and it isn't provided) though, it is now fixed.

Reply:
8818797c031a7e51ec3c56e1bf2e195d54ff6dee